File: 2015/Go/day-03/solution.go
package main;

import (
  "fmt"
  "io/ioutil"
)

type Coord struct {
  x int8;
  y int8;
}

func doP1(data string) uint16 {
  var visited = map[Coord]bool{};
  var c Coord;
  c.x, c.y = 0, 0;
  visited[c] = true;

  for i := 0; i < len(data); i++ {
    var d = rune(data[i]);
    if d == '^' {
      c.y++;
    } else if d == '>' {
      c.x++;
    } else if d == 'v' {
      c.y--;
    } else if d == '<' {
      c.x--;
    } else {
      continue;
    }
    visited[c] = true;
  }

  return uint16(len(visited));
}

func doP2(data string) uint16 {
  var visited = map[Coord]bool{};
  var s Coord;
  var r Coord;
  s.x, s.y = 0, 0;
  r.x, r.y = 0, 0;
  visited[s] = true;

  for i := 0; i < len(data); i++ {
    var d = rune(data[i]);
    var dx int8 = 0;
    var dy int8 = 0;
    if d == '^' {
      dy++;
    } else if d =='>' {
      dx++;
    } else if d == 'v' {
      dy--;
    } else if d == '<' {
      dx--;
    } else {
      continue;
    }

    if i % 2 == 0 {
      s.x += dx;
      s.y += dy;
      visited[s] = true;
    } else {
      r.x += dx;
      r.y += dy;
      visited[r] = true;
    }
  }
  return uint16(len(visited));
}

func main() {
  body, err := ioutil.ReadFile("input.txt");
  if err != nil {
    panic("Could not open input file.");
  }
  var data string = string(body);
  p1 := doP1(data);
  fmt.Printf("Day 03, Part 01: %d\n", p1);
  p2 := doP2(data);
  fmt.Printf("Day 03, Part 02: %d\n", p2);
}

File: 2015/Python/day-03/solution.py
with open("input.txt", "r") as f:
    data = f.read().strip()

# --- Part 1 ---
x, y = 0, 0
houses = set()
houses.add((x, y))

for d in list(data):
    if d == "^":
        y += 1
    elif d == ">":
        x += 1
    elif d == "v":
        y -= 1
    elif d == "<":
        x -= 1
    houses.add((x, y))
print("D03P01:", len(houses))

# --- Part 2 ---
sx, sy = 0, 0
rx, ry = 0, 0
houses.clear()
houses.add((0, 0))

for i, d in enumerate(list(data)):
    mod_x, mod_y = 0, 0
    if d == "^":
        mod_y += 1
    elif d == ">":
        mod_x += 1
    elif d == "v":
        mod_y -= 1
    elif d == "<":
        mod_x -= 1
    if i % 2 == 0:
        sx += mod_x
        sy += mod_y
        houses.add((sx, sy))
    else:
        rx += mod_x
        ry += mod_y
        houses.add((rx, ry))
print("D03P02:", len(houses))
